pallets/avn-anchor/src/tests.rs

```diff
@@ -614,13 +600,164 @@ fn checkpoint_id_overflow_fails() {
     });
 }
+
+// Fees
+#[test]
+fn set_checkpoint_fee_works() {
+    new_test_ext().execute_with(|| {
+        let chain_id = 0;
+        let new_fee = 100;
+
+        assert_ok!(AvnAnchor::set_checkpoint_fee(RuntimeOrigin::root(), chain_id, new_fee));
+
+        assert_eq!(AvnAnchor::checkpoint_fee(chain_id), new_fee);
+        System::assert_last_event(Event::CheckpointFeeUpdated { chain_id, new_fee }.into());
+    });
+}
+
+#[test]
+fn set_checkpoint_fee_fails_for_non_root() {
+    new_test_ext().execute_with(|| {
+        let non_root = create_account_id(1);
+        let chain_id = 0;
+        let new_fee = 100;
+
+        assert_noop!(
+            AvnAnchor::set_checkpoint_fee(RuntimeOrigin::signed(non_root), chain_id, new_fee),
+            DispatchError::BadOrigin
+        );
+    });
+}
+
+#[test]
+fn charge_fee_works() {
+    new_test_ext().execute_with(|| {
+        let handler = create_account_id(1);
+        let chain_id = 0;
+        let fee = 100;
+
+        assert_ok!(AvnAnchor::set_checkpoint_fee(RuntimeOrigin::root(), chain_id, fee));
+
+        assert_ok!(AvnAnchor::charge_fee(handler.clone(), chain_id));
+
+        System::assert_last_event(
+            Event::CheckpointFeeCharged { handler: handler.clone(), chain_id, fee }.into(),
+        );
+    });
+}
+
+#[test]
+fn submit_checkpoint_charges_correct_fee() {
+    new_test_ext().execute_with(|| {
+        let handler = create_account_id(1);
+        let chain_id = 0;
+        let fee = 100;
+        let name = bounded_vec(b"Test Chain");
+        let checkpoint = H256::random();
+
+        assert_ok!(AvnAnchor::register_chain_handler(RuntimeOrigin::signed(handler), name));
+        assert_ok!(AvnAnchor::set_checkpoint_fee(RuntimeOrigin::root(), chain_id, fee));
+
+        assert_ok!(AvnAnchor::submit_checkpoint_with_identity(
+            RuntimeOrigin::signed(handler),
+            checkpoint
+        ));
+
+        System::assert_has_event(
+            Event::CheckpointSubmitted(handler, chain_id, 0, checkpoint).into(),
+        );
+        System::assert_has_event(Event::CheckpointFeeCharged { handler, chain_id, fee }.into());
+    });
+}
+
+#[test]
+fn submit_checkpoint_charges_zero_fee() {
+    new_test_ext().execute_with(|| {
+        let handler = create_account_id(1);
+        let chain_id = 0;
+        let fee = 0;
+        let name = bounded_vec(b"Test Chain");
+        let checkpoint = H256::random();
+
+        assert_ok!(AvnAnchor::register_chain_handler(RuntimeOrigin::signed(handler), name));
+        assert_ok!(AvnAnchor::set_checkpoint_fee(RuntimeOrigin::root(), chain_id, fee));
+
+        assert_ok!(AvnAnchor::submit_checkpoint_with_identity(
+            RuntimeOrigin::signed(handler),
+            checkpoint
+        ));
+
+        System::assert_has_event(
+            Event::CheckpointSubmitted(handler, chain_id, 0, checkpoint).into(),
+        );
+        System::assert_has_event(Event::CheckpointFeeCharged { handler, chain_id, fee }.into());
```

Comment:
shouldn't we assert that the balance of the user was unchanged?
follow up:
zero fees meaning, not even gas fees or 0 check-pointing fee?